---
title: CisTopic on 10X RNA/ATAC
layout: cedar_analysis
author: Ryan Mulqueen
permalink: /cistopic_10xatac/
category: CEDAR
---

Running cisTopic on 10X Genomic ATAC/RNA Coassay (just the ATAC part). Taking Suerat Objects generated by Aaron Doe.

## File Location
```bash
/home/groups/CEDAR/doe/projects/ATACRNA/MCF7_justATAC_SO.rds
/home/groups/CEDAR/doe/projects/ATACRNA/T47D_justATAC_SO.rds
```

## Cistopic on ATAC data

```R
setwd("/home/groups/CEDAR/mulqueen/projects/10x_atacrna")
library(Signac)
library(Seurat)
library(SeuratWrappers)
library(cisTopic)
library(patchwork)
set.seed(1234)
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)

mcf7<-readRDS("/home/groups/CEDAR/doe/projects/ATACRNA/MCF7_justATAC_SO.rds") #generated by AD
t47d<-readRDS("/home/groups/CEDAR/doe/projects/ATACRNA/T47D_justATAC_SO.rds") #generated by AD

combined<- merge(mcf7, y = t47d, add.cell.ids = c("mcf7", "t47d"), project = "atac_rna")
combined<-subset(
  x = combined,
  subset = nCount_peaks > 3000 &
  nFeature_peaks > 10000 &
  TSS.enrichment > 4
)
saveRDS(combined,"210924_cellline.SeuratObject.Rds")
combined<-readRDS("210924_cellline.SeuratObject.Rds")

cistopic_generation<-function(x,name_out,outdir){
  wd<-outdir
  outname<-name_out
  atac_sub<-readRDS(x)
  cistopic_counts_frmt<-atac_sub$peaks@counts
  row.names(cistopic_counts_frmt)<-sub("-", ":", row.names(cistopic_counts_frmt))
  sub_cistopic<-cisTopic::createcisTopicObject(cistopic_counts_frmt)
  print("made cistopic object")
  sub_cistopic_models<-cisTopic::runWarpLDAModels(sub_cistopic,topic=c(10:30),nCores=7,addModels=FALSE)
  saveRDS(sub_cistopic_models,file=paste0(wd,"/",outname,".CisTopicObject.Rds"))
  print("finshed running cistopic")

  pdf(paste0(wd,"/",outname,"_model_selection.pdf"))
  par(mfrow=c(3,3))
  sub_cistopic_models<- selectModel(sub_cistopic_models, type='derivative')
  dev.off()

  #run UMAP on topics
  topic_df<-as.data.frame(sub_cistopic_models@selected.model$document_expects)
  row.names(topic_df)<-paste0("Topic_",row.names(topic_df))
  dims<-as.data.frame(uwot::umap(t(topic_df),n_components=2))
  row.names(dims)<-colnames(topic_df)
  colnames(dims)<-c("x","y")
  dims$cellID<-row.names(dims)
  dims<-merge(dims,atac_sub@meta.data,by.x="cellID",by.y="row.names")


  #Add cell embeddings into seurat
  cell_embeddings<-as.data.frame(sub_cistopic_models@selected.model$document_expects)
  colnames(cell_embeddings)<-sub_cistopic_models@cell.names
  n_topics<-nrow(cell_embeddings)
  row.names(cell_embeddings)<-paste0("topic_",1:n_topics)
  cell_embeddings<-as.data.frame(t(cell_embeddings))

  #Add feature loadings into seurat
  feature_loadings<-as.data.frame(sub_cistopic_models@selected.model$topics)
  row.names(feature_loadings)<-paste0("topic_",1:n_topics)
  feature_loadings<-as.data.frame(t(feature_loadings))

  #combined cistopic results (cistopic loadings and umap with seurat object)
  cistopic_obj<-CreateDimReducObject(embeddings=as.matrix(cell_embeddings),loadings=as.matrix(feature_loadings),assay="peaks",key="topic_")
  umap_dims<-as.data.frame(as.matrix(dims[2:3]))
  colnames(umap_dims)<-c("UMAP_1","UMAP_2")
  row.names(umap_dims)<-dims$cellID
  cistopic_umap<-CreateDimReducObject(embeddings=as.matrix(umap_dims),assay="peaks",key="UMAP_")
  atac_sub@reductions$cistopic<-cistopic_obj
  atac_sub@reductions$umap<-cistopic_umap

  n_topics<-ncol(Embeddings(atac_sub,reduction="cistopic"))

  atac_sub <- FindNeighbors(object = atac_sub, reduction = 'cistopic', dims = 1:n_topics ) 
  atac_sub <- FindClusters(object = atac_sub, verbose = TRUE, graph.name="peaks_snn", resolution=0.2 ) 
  plt1<-DimPlot(atac_sub,group.by=c("origin","seurat_clusters"))
  plt2<-FeaturePlot(atac_sub,features=c("nucleosome_signal","TSS.enrichment","nCount_peaks","nFeature_peaks"))
  pdf(paste0(outname,".umap.pdf"))
  plt1
  plt2
  dev.off()
  system(paste0("slack -F ",outname,".umap.pdf"," ryan_todo"))
  saveRDS(atac_sub,paste0(outname,".SeuratObject.Rds"))

  }

cistopic_generation(x="210924_cellline.SeuratObject.Rds",name_out="210924_cellline",outdir="/home/groups/CEDAR/mulqueen/projects/10x_atacrna")

#Expanding on cistopic object analysis
obj<-readRDS("210924_cellline.SeuratObject.Rds")
cisTopicObject<-readRDS("/home/groups/CEDAR/mulqueen/projects/10x_atacrna/210924_cellline.CisTopicObject.Rds")
cisTopicObject <- addCellMetadata(cisTopicObject, cell.data = obj@meta.data)

#heatmap by topic
cisTopicObject<- selectModel(cisTopicObject, type='derivative')
cisTopicObject <- runUmap(cisTopicObject, target='cell')
pdf("celltopic_heatmap.pdf")
cellTopicHeatmap(cisTopicObject, method='Probability', colorBy=c("origin","seurat_clusters"))
dev.off()
system("slack -F celltopic_heatmap.pdf ryan_todo")

#region score association
pred.matrix <- predictiveDistribution(cisTopicObject)
cisTopicObject <- binarizecisTopics(cisTopicObject, thrP=0.999, plot=TRUE)
cisTopicObject <- getRegionsScores(cisTopicObject, method='NormTop', scale=TRUE)

txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
cisTopicObject <- annotateRegions(cisTopicObject, txdb=TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb='org.Hs.eg.db')
cisTopicObject <- runtSNE(cisTopicObject, target='region', perplexity=200, check_duplicates=FALSE) #cluster by regions

getBedFiles(cisTopicObject, path='cisTopics_asBed')

saveRDS(cisTopicObject,"/home/groups/CEDAR/mulqueen/projects/10x_atacrna/210924_cellline.CisTopicObject.Rds")

#save cistopic region data
outdat<-lapply(cisTopicObject@binarized.cisTopics, function(x) {
  temp<-merge(x,cisTopicObject@region.data,by="row.names")
  temp$topic_assigned<-names(x)
  return(temp)})

outdat<-data.table::rbindlist(outdat,use.names=FALSE)
write.table(outdat,file="topic_region_assignment.txt",col.names=T,row.names=T,quote=F,sep="\t")
system("slack -F topic_region_assignment.txt ryan_todo" )
pdf("topic_annotations.pdf")
par(mfrow=c(1,1))
signaturesHeatmap(cisTopicObject, selected.signatures = 'annotation')
plotFeatures(cisTopicObject, method='tSNE', target='region', topic_contr=NULL, colorBy=c('annotation'), cex.legend = 0.8, factor.max=.75, dim=2, legend=TRUE, intervals=20)
dev.off()
system("slack -F topic_annotations.pdf ryan_todo")


pdf("topic_scores.pdf")
par(mfrow=c(4,4))
plotFeatures(cisTopicObject, method='Umap', target='cell', topic_contr='Probability', colorBy=NULL, cex.legend = 0.8, factor.max=.75, dim=2, legend=TRUE)
dev.off()
system("slack -F topic_scores.pdf ryan_todo")

#print foxm1 associated topics
unlist(lapply(cisTopicObject@binarized.cisTopics,function(x) {
  print(
    paste(
      names(x),
      sum(row.names(cisTopicObject@region.data[cisTopicObject@region.data$SYMBOL %in% c("FOXM1"),]) %in% row.names(x))
      ))}))
#8 and 18
```

### ChromVar for Transcription Factor Motifs


```R
  library(Signac)
  library(Seurat)
  library(JASPAR2020)
  library(TFBSTools)
  library(BSgenome.Hsapiens.UCSC.hg19)
  library(patchwork)
  set.seed(1234)
  library(BiocParallel)
  register(MulticoreParam(5))

setwd("/home/groups/CEDAR/mulqueen/projects/10x_atacrna")

combined<-readRDS("210924_cellline.SeuratObject.Rds")


  #Read in data and modify to monocle CDS file
  #read in RDS file.

  # Get a list of motif position frequency matrices from the JASPAR database
  pfm <- getMatrixSet(
    x = JASPAR2020,
    opts = list(species =9606, all_versions = FALSE))

  # Scan the DNA sequence of each peak for the presence of each motif, using orgo_atac for all objects (shared peaks)
  motif.matrix.hg19 <- CreateMotifMatrix(
    features = granges(combined[["peaks"]]),
    pwm = pfm,
    genome = 'hg19',
    use.counts = FALSE)

  # Create a new Mofif object to store the results
  motif.hg19 <- CreateMotifObject(
    data = motif.matrix.hg19,
    pwm = pfm)

  #for human
   combined <- SetAssayData(
    object = combined,
    assay = 'peaks',
    slot = 'motifs',
    new.data = motif.hg19)

  combined <- RegionStats(object = combined, genome = BSgenome.Hsapiens.UCSC.hg19,assay="peaks")
  combined <- RunChromVAR( object = combined,genome = BSgenome.Hsapiens.UCSC.hg19,assay="peaks")
  saveRDS(combined,file="210924_cellline.SeuratObject.Rds")


#define DA functions for parallelization
#Use LR test for atac data
da_one_v_rest<-function(i,obj,group,assay.="chromvar",latent.vars.="nCount_peaks"){
    da_ga_tmp <- FindMarkers(
        object = obj,
        ident.1 = i,
        group.by = group,
        test.use = 'LR',
        latent.vars = latent.vars.,
        only.pos=T,
        assay=assay.
        )
    da_ga_tmp$da_region<-row.names(da_ga_tmp)
    da_ga_tmp$enriched_group<-c(i)
    da_ga_tmp$compared_group<-c("all_other_cells")
    return(da_ga_tmp)
  }


n.cores=10 #Perform parallel application of DA test
da_ga<-mclapply(
    unique(combined$origin),
    FUN=da_one_v_rest,
    obj=combined,
    group="origin", 
    assay.="chromvar",
    mc.cores=n.cores)

da_ga_df<-do.call("rbind",da_ga) #Merge the final data frame from the list for 1vrest DA
da_ga_df$tf_name <- unlist(lapply(unlist(lapply(da_ga_df$da_region, function(x) getMatrixByID(JASPAR2020,ID=x))),function(y) name(y)))
write.table(da_ga_df,file="210924_cellline.onevrest.da_chromvar.txt",sep="\t",col.names=T,row.names=T,quote=F)

library(dplyr)
for (i in unique(combined$origin)){print(as.data.frame(da_ga_df %>% filter(enriched_group==i) %>% dplyr::arrange(p_val_adj) %>% dplyr::slice(1:10)))}

n.cores=10 #Perform parallel application of DA test
da_ga<-mclapply(
    unique(combined$origin),
    FUN=da_one_v_rest,
    obj=combined,
    group="origin", 
    assay.="peaks",
    mc.cores=n.cores)

da_ga_df<-do.call("rbind",da_ga) #Merge the final data frame from the list for 1vrest DA
write.table(da_ga_df,file="210924_cellline.onevrest.da_peaks.txt",sep="\t",col.names=T,row.names=T,quote=F)



########################Plot out top TF for each cluster###################
library(dplyr)
library(ComplexHeatmap)
library(circlize)
  library(viridis)
da_tf<-read.csv(file="210924_cellline.onevrest.da_chromvar.txt",head=T,sep="\t",row.names=NULL)
da_tf<-da_tf[complete.cases(da_tf),]

da_tf$label<-""
for (x in unique(da_tf$enriched_group)){
selc_genes<-as.data.frame(da_tf %>% filter(enriched_group==x)  %>% filter(p_val < 0.05) %>% arrange(rev(desc(p_val_adj))) %>% slice(1:10))$tf_name
da_tf[da_tf$tf_name %in% selc_genes & da_tf$enriched_group==x,]$label<- da_tf[da_tf$tf_name %in% selc_genes & da_tf$enriched_group==x,]$tf_name
}

#Get gene activity scores data frame to summarize over subclusters (limit to handful of marker genes)
dat_tf<-as.data.frame(t(as.data.frame(combined[["chromvar"]]@data)))
sum_tf<-split(dat_tf,combined$origin) #group by rows to seurat clusters
sum_tf<-lapply(sum_tf,function(x) apply(x,2,mean)) #take average across group
sum_tf<-do.call("rbind",sum_tf) #condense to smaller data frame

sum_tf<-t(scale(sum_tf))
sum_tf<-sum_tf[,!endsWith(colnames(sum_tf),"NA")] #remove NA (doublet cells)

sum_tf<-sum_tf[row.names(sum_tf) %in% unique(da_tf[da_tf$label!="",]$da_region),]
row.names(sum_tf)<-da_tf[match(row.names(sum_tf),da_tf$da_region,nomatch=0),]$tf_name
sum_tf_plot<-na.omit(t(sum_tf))

colfun=colorRamp2(quantile(unlist(sum_tf_plot), probs=c(0.5,0.75,0.95)),cividis(3))
plt1<-Heatmap(sum_tf_plot,
    #cluster_rows=sum_da_dend,
    #left_annotation=side_ha,
    col=colfun,
    #bottom_annotation=bottom_ha,
    column_names_gp = gpar(fontsize = 14),
    row_names_gp=gpar(fontsize=14),
    column_names_rot=90
)

plt1<-draw(plt1)


pdf("combined.tf.heatmap.pdf",height=20,width=20)
draw(plt1)
dev.off()
system("slack -F combined.tf.heatmap.pdf ryan_todo")


```


<!---RUN CICERO-->
<!--RUN cistopic by cell line-->
<!--Check topics from TITAN against cistopic-->
<!-- chipseq foxm1 and esr1 overlap of topics-->
<!-- regress out nCount_peaks from umap during scaling-->
<!-- get more stringent peak list from topics-->