---
title: CisTopic on 10X RNA/ATAC
layout: cedar_analysis
author: Ryan Mulqueen
permalink: /cistopic_10xatac/
category: CEDAR
---

Running cisTopic on 10X Genomic ATAC/RNA Coassay (just the ATAC part). Taking Suerat Objects generated by Aaron Doe.

## File Location
```bash
/home/groups/CEDAR/doe/projects/ATACRNA/MCF7_justATAC_SO.rds
/home/groups/CEDAR/doe/projects/ATACRNA/T47D_justATAC_SO.rds
```

## Reference data
Using chipseq bed files held on cistrome for accessibility analysis.

```bash
/home/groups/CEDAR/mulqueen/ref/cistrome #stored reference files here, downloaded from site at .tar.gz file
/home/groups/CEDAR/mulqueen/ref/cistrome/human_factor_full_QC.txt #has information on each download peaks files
/home/groups/CEDAR/mulqueen/ref/cistrome/human_factor #has individual peak files
```

## Cistopic on ATAC data

```R
setwd("/home/groups/CEDAR/mulqueen/projects/10x_atacrna")
library(Signac)
library(Seurat)
library(SeuratWrappers)
library(cisTopic)
library(patchwork)
set.seed(1234)
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(AUCell)
library(rtracklayer)
library(parallel)

#read in RNA data to subset same cells
mcf7_rna<-readRDS("/home/groups/CEDAR/doe/projects/ATACRNA/MCF7_RNA_strictQC.rds")
t47d_rna<-readRDS("/home/groups/CEDAR/doe/projects/ATACRNA/T47D_RNA_strictQC.rds")
mcf7_rna<-RenameCells(mcf7_rna,new.names=paste0(row.names(mcf7_rna@meta.data),mcf7_rna$origin))
t47d_rna<-RenameCells(t47d_rna,new.names=paste0(row.names(t47d_rna@meta.data),t47d_rna$origin))

#read in pipeline generated atac data
mcf7<-readRDS("/home/groups/CEDAR/doe/projects/ATACRNA/MCF7_justATAC_SO.rds") #generated by AD
t47d<-readRDS("/home/groups/CEDAR/doe/projects/ATACRNA/T47D_justATAC_SO.rds") #generated by AD

#renaming cells to fit atac data, have to add origin so cell names are unique
mcf7<-RenameCells(mcf7,new.names=paste0(unlist(lapply(strsplit(row.names(mcf7@meta.data),"_"),"[",1)),mcf7$origin))
t47d<-RenameCells(t47d,new.names=paste0(unlist(lapply(strsplit(row.names(t47d@meta.data),"_"),"[",1)),t47d$origin))

#subset cells
mcf7<-subset(x=mcf7,cells=row.names(mcf7_rna@meta.data))
t47d<-subset(x=t47d,cells=row.names(t47d_rna@meta.data))

combined<- merge(mcf7, y = t47d, add.cell.ids = c("mcf7", "t47d"), project = "atac_rnasubsetted")

#a priori filter, not using to stay consistent with rna data
#combined<-subset(
#  x = combined,
#  subset = nCount_peaks > 3000 &
#  nFeature_peaks > 10000 &
#  TSS.enrichment > 4
#)

saveRDS(combined,"210924_cellline.SeuratObject.Rds")
saveRDS(mcf7,"210924_mcf7.SeuratObject.Rds")
saveRDS(t47d,"210924_t47d.SeuratObject.Rds")

cistopic_generation<-function(x,name_out,outdir="/home/groups/CEDAR/mulqueen/projects/10x_atacrna"){
  wd<-outdir
  outname<-name_out
  atac_sub<-x
  cistopic_counts_frmt<-atac_sub$peaks@counts
  row.names(cistopic_counts_frmt)<-sub("-", ":", row.names(cistopic_counts_frmt))
  sub_cistopic<-cisTopic::createcisTopicObject(cistopic_counts_frmt)
  print("made cistopic object")
  sub_cistopic_models<-cisTopic::runWarpLDAModels(sub_cistopic,topic=c(10:30),nCores=7,addModels=FALSE)
  sub_cistopic_models<-addCellMetadata(sub_cistopic_models, cell.data =x@meta.data)
  pdf(paste0(wd,"/",outname,"_model_selection.pdf"))
  par(mfrow=c(3,3))
  sub_cistopic_models<- selectModel(sub_cistopic_models, type='derivative')
  dev.off()
  system(paste0("slack -F ",paste0(wd,"/",outname,"_model_selection.pdf")," ryan_todo"))
  
  saveRDS(sub_cistopic_models,file=paste0(wd,"/",outname,".CisTopicObject.Rds"))
  sub_cistopic_models<-readRDS(file=paste0(wd,"/",outname,".CisTopicObject.Rds"))
  print("finshed running cistopic")

  #Add cell embeddings into seurat
  cell_embeddings<-as.data.frame(sub_cistopic_models@selected.model$document_expects)
  colnames(cell_embeddings)<-sub_cistopic_models@cell.names
  n_topics<-nrow(cell_embeddings)
  row.names(cell_embeddings)<-paste0("topic_",1:n_topics)
  cell_embeddings<-as.data.frame(t(cell_embeddings))

  #Add feature loadings into seurat
  feature_loadings<-as.data.frame(sub_cistopic_models@selected.model$topics)
  row.names(feature_loadings)<-paste0("topic_",1:n_topics)
  feature_loadings<-as.data.frame(t(feature_loadings))

  #combined cistopic results (cistopic loadings and umap with seurat object)
  cistopic_obj<-CreateDimReducObject(embeddings=as.matrix(cell_embeddings),loadings=as.matrix(feature_loadings),assay="peaks",key="topic_")
  atac_sub@reductions$cistopic<-cistopic_obj
  n_topics<-ncol(Embeddings(atac_sub,reduction="cistopic")) #add scaling for ncount peaks somewhere in here
  atac_sub<-RunUMAP(atac_sub,reduction="cistopic",dims=1:n_topics)
  atac_sub <- FindNeighbors(object = atac_sub, reduction = 'cistopic', dims = 1:n_topics ) 
  atac_sub <- FindClusters(object = atac_sub, verbose = TRUE, graph.name="peaks_snn", resolution=0.2 ) 
  plt1<-DimPlot(atac_sub,reduction="umap",group.by=c("origin","seurat_clusters"))
  plt2<-FeaturePlot(atac_sub,reduction="umap",features=c("nucleosome_signal","TSS.enrichment","nCount_peaks","nFeature_peaks"))
  pdf(paste0(wd,"/",outname,".umap.pdf"))
  print(plt1)
  print(plt2)
  dev.off()
  system(paste0("slack -F ",paste0(wd,"/",outname,".umap.pdf")," ryan_todo"))
  saveRDS(atac_sub,paste0(wd,"/",outname,".SeuratObject.Rds"))
  }


#Running cistopic
  combined<-readRDS("210924_cellline.SeuratObject.Rds")
  cistopic_generation(x=combined,name_out="210924_cellline")

  #Run just for t47D
  t47d<-readRDS("210924_t47d.SeuratObject.Rds")
  cistopic_generation(x=t47d,name_out="210924_t47d")

  #Run just for mcf7
  mcf7<-readRDS("210924_mcf7.SeuratObject.Rds")
  cistopic_generation(x=mcf7,name_out="210924_mcf7")

```

Take cistopic data and perform bed file overlap using cistrome cell line specific ChIP-seq.


```R
setwd("/home/groups/CEDAR/mulqueen/projects/10x_atacrna")
library(Signac)
library(Seurat)
library(SeuratWrappers)
library(cisTopic)
library(patchwork)
set.seed(1234)
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(AUCell)
library(rtracklayer)
library(parallel)
library(ComplexHeatmap)

#read in cistrome data for topic analysis
cistrome_db<-read.csv("/home/groups/CEDAR/mulqueen/ref/cistrome/human_factor_full_QC.txt",sep="\t") #has information on each download peaks files
cistrome_db<-cistrome_db[cistrome_db$Cell_line %in% c("MCF-7","T47D"),] #limit to our cell types
cistrome_db<-cistrome_db[cistrome_db$PeaksFoldChangeAbove10>1000,] #set lower limit for peaks
cistrome_dir="/home/groups/CEDAR/mulqueen/ref/cistrome/human_factor" #has individual peak files
ChIP_Seq_signatures <- paste(cistrome_dir, list.files(cistrome_dir), sep='/') #grab all files in directory
signature_dcids<-unlist(lapply(strsplit(unlist(lapply(strsplit(ChIP_Seq_signatures,"human_factor/"),"[",2)),"_sort"),"[",1)) #grab dcID from file name
ChIP_Seq_signatures <- ChIP_Seq_signatures[signature_dcids %in% cistrome_db$DCid] #limit to those in cistrome_db after filter
ChIP_Seq_signatures<-ChIP_Seq_signatures[order(as.numeric(unlist(lapply(strsplit(unlist(lapply(strsplit(ChIP_Seq_signatures,"human_factor/"),"[",2)),"_sort"),"[",1))))] #sort by numeric, like db
cistrome_db$label<-paste(cistrome_db$Cell_line,cistrome_db$Factor,cistrome_db$DCid,sep="_")
cistrome_db$file<-ChIP_Seq_signatures

#need to liftover files since cistrome is hg38 and this data is hg19, following https://www.bioconductor.org/packages/release/workflows/vignettes/liftOver/inst/doc/liftov.html
path = system.file(package="liftOver", "extdata", "hg38ToHg19.over.chain")
ch = import.chain(path)

liftover_bed<-function(x){
  label<-cistrome_db[cistrome_db$file==x,]$label
  print(paste(x,label))
  x_hg38<-read.table(x,sep="\t")
  colnames(x_hg38)[1:3]<-c("chr","start","end")  
  x_hg38<-x_hg38[x_hg38$chr %in% paste0("chr",c(1:23,"X","Y")),]
  x_hg38<-makeGRangesFromDataFrame(x_hg38) #read in bed file and make granges
  seqlevelsStyle(x_hg38) = "UCSC"  # necessary
  x_hg19<-liftOver(x_hg38, ch)
  x_hg19<-unlist(x_hg19)
  genome(x_hg19) = "hg19"
  export.bed(object=x_hg19,con=paste(cistrome_dir,paste0(label,".hg19liftOver.bed"),sep="/"),format="bed")
}

mclapply(as.character(cistrome_db$file),liftover_bed,mc.cores=10) #parallelize chipseq to 20 cores, saving bed files after liftover with new name
hg19_chip<-paste(cistrome_dir, list.files(cistrome_dir,pattern="hg19liftOver.bed"), sep='/') #grab all files in directory
labels<-unlist(lapply(strsplit(unlist(lapply(strsplit(hg19_chip,"human_factor/"),"[",2)),".hg19liftOver"),"[",1)) #grab labels from file names


#run through further processing
#this function will perform more analysis on topics in the selected model, looking at 
#1. cell topic weights
#2. cistrome chipseq peak overlaps per topic
#3. topic annotations (if peaks fall primarily in promoter regions etc)

cistopic_processing<-function(x,y,outname){
  obj<-x
  cisTopicObject<-y
  cisTopicObject<-addCellMetadata(cisTopicObject, cell.data =obj@meta.data)
  #heatmap by topic
  cisTopicObject <- runUmap(cisTopicObject, target='cell')
  pdf(paste(outname,"celltopic_heatmap.pdf",sep="."))
  print(cellTopicHeatmap(cisTopicObject, method='Probability', colorBy=c("origin","seurat_clusters")))
  dev.off()
  system(paste("slack -F", paste(outname,"celltopic_heatmap.pdf",sep="."), "ryan_todo",sep=" "))

  #region score association
  pred.matrix <- predictiveDistribution(cisTopicObject)
  cisTopicObject <- getSignaturesRegions(cisTopicObject, hg19_chip, labels=labels, minOverlap = 0.01) #output new bed files to be read in and processed
  aucellRankings <- AUCell_buildRankings(pred.matrix, plot=FALSE, verbose=FALSE)   # Compute cell rankings
  cisTopicObject <- signatureCellEnrichment(cisTopicObject, aucellRankings, selected.signatures='all', aucMaxRank = 0.1*nrow(aucellRankings),nCores=1,plot=FALSE)   # Check signature enrichment in cells

  cisTopicObject <- getRegionsScores(cisTopicObject, method='NormTop', scale=TRUE)
  cisTopicObject <- binarizecisTopics(cisTopicObject, thrP=0.999, plot=TRUE)


  pdf(paste(outname,"signature_heatmap.pdf",sep="."),height=40)
  signaturesHeatmap(cisTopicObject,row_names_gp = gpar(fontsize = 3))
  dev.off()
  system(paste("slack -F", paste(outname,"signature_heatmap.pdf",sep="."), "ryan_todo",sep=" "))

  txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
  cisTopicObject <- annotateRegions(cisTopicObject, txdb=TxDb.Hsapiens.UCSC.hg19.knownGene, annoDb='org.Hs.eg.db')
  cisTopicObject <- runtSNE(cisTopicObject, target='region', perplexity=200, check_duplicates=FALSE) #cluster by regions
  saveRDS(cisTopicObject,paste0(outname,".CisTopicObject.Rds"))

  #save cistopic region data
  outdat<-lapply(cisTopicObject@binarized.cisTopics, function(x) {
    temp<-merge(x,cisTopicObject@region.data,by="row.names")
    temp$topic_assigned<-names(x)
    return(temp)})

  outdat<-data.table::rbindlist(outdat,use.names=FALSE)
  write.table(outdat,file=paste0(outname,".topic_region_assignment.txt"),col.names=T,row.names=T,quote=F,sep="\t")
  system(paste("slack -F ",paste0(outname,".topic_region_assignment.txt")," ryan_todo" ))
  pdf(paste0(outname,".topic_annotations.pdf"))
  par(mfrow=c(1,1))
  signaturesHeatmap(cisTopicObject, selected.signatures = 'annotation')
  plotFeatures(cisTopicObject, method='tSNE', target='region', topic_contr=NULL, colorBy=c('annotation'), cex.legend = 0.8, factor.max=.75, dim=2, legend=TRUE, intervals=20)
  dev.off()
  system(paste("slack -F ",paste0(outname,".topic_annotations.pdf")," ryan_todo" ))

  pdf(paste0(outname,".topic_scores.pdf"))
  par(mfrow=c(4,4))
  plotFeatures(cisTopicObject, method='Umap', target='cell', topic_contr='Probability', colorBy=NULL, cex.legend = 0.8, factor.max=.75, dim=2, legend=TRUE)
  dev.off()
  system(paste("slack -F ",paste0(outname,".topic_scores.pdf")," ryan_todo" ))
}



#add cistopic function to all cells
  combined<-readRDS("210924_cellline.SeuratObject.Rds")
  combined_cisTopicObject<-readRDS("/home/groups/CEDAR/mulqueen/projects/10x_atacrna/210924_cellline.CisTopicObject.Rds")
  cistopic_processing(x=combined,y=combined_cisTopicObject,outname="210924_cellline")

  t47d<-readRDS("210924_t47d.SeuratObject.Rds")
  t47d_cisTopicObject<-readRDS("/home/groups/CEDAR/mulqueen/projects/10x_atacrna/210924_t47d.CisTopicObject.Rds")
  cistopic_processing(x=t47d,y=t47d_cisTopicObject,outname="210924_t47d")

  mcf7<-readRDS("210924_mcf7.SeuratObject.Rds")
  mcf7_cisTopicObject<-readRDS("/home/groups/CEDAR/mulqueen/projects/10x_atacrna/210924_mcf7.CisTopicObject.Rds")
  cistopic_processing(x=mcf7,y=mcf7_cisTopicObject,outname="210924_mcf7")

```

### ChromVar for Transcription Factor Motifs

```R
  library(Signac)
  library(Seurat)
  library(JASPAR2020)
  library(TFBSTools)
  library(BSgenome.Hsapiens.UCSC.hg19)
  library(patchwork)
  set.seed(1234)
  library(BiocParallel)
  register(MulticoreParam(5))

setwd("/home/groups/CEDAR/mulqueen/projects/10x_atacrna")

combined<-readRDS("210924_cellline.SeuratObject.Rds")
mcf7<-readRDS("210924_mcf7.SeuratObject.Rds")
t47d<-readRDS("210924_t47d.SeuratObject.Rds")


  #Read in data and modify to monocle CDS file
  #read in RDS file.

  # Get a list of motif position frequency matrices from the JASPAR database
  pfm <- getMatrixSet(
    x = JASPAR2020,
    opts = list(species =9606, all_versions = FALSE))

  # Scan the DNA sequence of each peak for the presence of each motif, using orgo_atac for all objects (shared peaks)
  # Create a new Mofif object to store the results
  #for combined
    motif.matrix.hg19 <- CreateMotifMatrix(features = granges(combined[["peaks"]]), pwm = pfm, genome = 'hg19', use.counts = FALSE)
    motif.hg19 <- CreateMotifObject(data = motif.matrix.hg19, pwm = pfm)
  combined <- SetAssayData(object = combined, assay = 'peaks', slot = 'motifs', new.data = motif.hg19)
  combined <- RegionStats(object = combined, genome = BSgenome.Hsapiens.UCSC.hg19,assay="peaks")
  combined <- RunChromVAR( object = combined,genome = BSgenome.Hsapiens.UCSC.hg19,assay="peaks")
  saveRDS(combined,file="210924_cellline.SeuratObject.Rds")

    motif.matrix.hg19 <- CreateMotifMatrix(features = granges(mcf7[["peaks"]]), pwm = pfm, genome = 'hg19', use.counts = FALSE)
    motif.hg19 <- CreateMotifObject(data = motif.matrix.hg19, pwm = pfm)
  mcf7 <- SetAssayData(object = mcf7, assay = 'peaks', slot = 'motifs', new.data = motif.hg19)
  mcf7 <- RegionStats(object = mcf7, genome = BSgenome.Hsapiens.UCSC.hg19,assay="peaks")
  mcf7 <- RunChromVAR( object = mcf7,genome = BSgenome.Hsapiens.UCSC.hg19,assay="peaks")
  saveRDS(mcf7,file="210924_mcf7.SeuratObject.Rds")

    motif.matrix.hg19 <- CreateMotifMatrix(features = granges(t47d[["peaks"]]), pwm = pfm, genome = 'hg19', use.counts = FALSE)
    motif.hg19 <- CreateMotifObject(data = motif.matrix.hg19, pwm = pfm)
  t47d <- SetAssayData(object = t47d, assay = 'peaks', slot = 'motifs', new.data = motif.hg19)
  t47d <- RegionStats(object = t47d, genome = BSgenome.Hsapiens.UCSC.hg19,assay="peaks")
  t47d <- RunChromVAR( object = t47d,genome = BSgenome.Hsapiens.UCSC.hg19,assay="peaks")
  saveRDS(t47d,file="210924_t47d.SeuratObject.Rds")

#define DA functions for parallelization
#Use LR test for atac data
da_one_v_rest<-function(i,obj,group,assay.="chromvar",latent.vars.="nCount_peaks"){
    da_ga_tmp <- FindMarkers(
        object = obj,
        ident.1 = i,
        group.by = group,
        test.use = 'LR',
        latent.vars = latent.vars.,
        only.pos=T,
        assay=assay.
        )
    da_ga_tmp$da_region<-row.names(da_ga_tmp)
    da_ga_tmp$enriched_group<-c(i)
    da_ga_tmp$compared_group<-c("all_other_cells")
    return(da_ga_tmp)
  }


#combined
n.cores=1 #Perform parallel application of DA test
da_ga<-mclapply(unique(combined$origin), FUN=da_one_v_rest, obj=combined, group="origin", assay.="chromvar", mc.cores=n.cores)
da_ga_df<-do.call("rbind",da_ga) #Merge the final data frame from the list for 1vrest DA
da_ga_df$tf_name <- unlist(lapply(unlist(lapply(da_ga_df$da_region, function(x) getMatrixByID(JASPAR2020,ID=x))),function(y) name(y)))
write.table(da_ga_df,file="210924_cellline.onevrest.da_chromvar.txt",sep="\t",col.names=T,row.names=T,quote=F)

library(dplyr)
for (i in unique(combined$origin)){print(as.data.frame(da_ga_df %>% filter(enriched_group==i) %>% dplyr::arrange(p_val_adj) %>% dplyr::slice(1:10)))}

n.cores=1 #Perform parallel application of DA test
da_ga<-mclapply(unique(combined$origin), FUN=da_one_v_rest, obj=combined, group="origin", assay.="peaks", mc.cores=n.cores)
da_ga_df<-do.call("rbind",da_ga) #Merge the final data frame from the list for 1vrest DA
write.table(da_ga_df,file="210924_cellline.onevrest.da_peaks.txt",sep="\t",col.names=T,row.names=T,quote=F)



########################Plot out top TF for each cluster###################
library(dplyr)
library(ComplexHeatmap)
library(circlize)
  library(viridis)
da_tf<-read.csv(file="210924_cellline.onevrest.da_chromvar.txt",head=T,sep="\t",row.names=NULL)
da_tf<-da_tf[complete.cases(da_tf),]

da_tf$label<-""
for (x in unique(da_tf$enriched_group)){
selc_genes<-as.data.frame(da_tf %>% filter(enriched_group==x)  %>% filter(p_val < 0.05) %>% arrange(rev(desc(p_val_adj))) %>% slice(1:10))$tf_name
da_tf[da_tf$tf_name %in% selc_genes & da_tf$enriched_group==x,]$label<- da_tf[da_tf$tf_name %in% selc_genes & da_tf$enriched_group==x,]$tf_name
}

#Get gene activity scores data frame to summarize over subclusters (limit to handful of marker genes)
dat_tf<-as.data.frame(t(as.data.frame(combined[["chromvar"]]@data)))
sum_tf<-split(dat_tf,combined$origin) #group by rows to seurat clusters
sum_tf<-lapply(sum_tf,function(x) apply(x,2,mean)) #take average across group
sum_tf<-do.call("rbind",sum_tf) #condense to smaller data frame

sum_tf<-t(scale(sum_tf))
sum_tf<-sum_tf[,!endsWith(colnames(sum_tf),"NA")] #remove NA (doublet cells)

sum_tf<-sum_tf[row.names(sum_tf) %in% unique(da_tf[da_tf$label!="",]$da_region),]
row.names(sum_tf)<-da_tf[match(row.names(sum_tf),da_tf$da_region,nomatch=0),]$tf_name
sum_tf_plot<-na.omit(t(sum_tf))

colfun=colorRamp2(quantile(unlist(sum_tf_plot), probs=c(0.5,0.75,0.95)),cividis(3))
plt1<-Heatmap(sum_tf_plot,
    #cluster_rows=sum_da_dend,
    #left_annotation=side_ha,
    col=colfun,
    #bottom_annotation=bottom_ha,
    column_names_gp = gpar(fontsize = 14),
    row_names_gp=gpar(fontsize=14),
    column_names_rot=90
)

plt1<-draw(plt1)


pdf("combined.tf.heatmap.pdf",height=20,width=20)
draw(plt1)
dev.off()
system("slack -F combined.tf.heatmap.pdf ryan_todo")


```

## Motif Footprinting
Based on https://satijalab.org/signac/articles/footprint.html
```R
  library(Signac)
  library(Seurat)
  library(JASPAR2020)
  library(TFBSTools)
  library(BSgenome.Hsapiens.UCSC.hg19)
  library(patchwork)
  set.seed(1234)
  library(motifmatchr)
  library(parallel)
  setwd("/home/groups/CEDAR/mulqueen/projects/10x_atacrna")

combined<-readRDS("210924_cellline.SeuratObject.Rds")
mcf7<-readRDS("210924_mcf7.SeuratObject.Rds")
t47d<-readRDS("210924_t47d.SeuratObject.Rds")

# extract position frequency matrices for the motifs
pwm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(species = 9606, all_versions = FALSE)
)

# add motif information
DefaultAssay(combined)<-"peaks"
combined <- AddMotifs(combined, genome = BSgenome.Hsapiens.UCSC.hg19, pfm = pwm)
DefaultAssay(mcf7)<-"peaks"
mcf7 <- AddMotifs(mcf7, genome = BSgenome.Hsapiens.UCSC.hg19, pfm = pwm)
DefaultAssay(t47d)<-"peaks"
t47d <- AddMotifs(t47d, genome = BSgenome.Hsapiens.UCSC.hg19, pfm = pwm)

saveRDS(combined,file="210924_cellline.SeuratObject.Rds")
saveRDS(mcf7,file="210924_mcf7.SeuratObject.Rds")
saveRDS(t47d,file="210924_t47d.SeuratObject.Rds")

#function for plotting footprints in data sets

plot_footprints<-function(x,footprints,outname){
  x <- Footprint(object = x, motif.name = footprints, genome = BSgenome.Hsapiens.UCSC.hg19, in.peaks=T)   # gather the footprinting information for sets of motifs
  p2 <- PlotFootprint(x, features = footprints,label=F)   # plot the footprint data for each group of cells, might want to change idents
  return(p2)
}

Idents(combined)<-combined$origin
out<-mclapply(c("GATA3","ESR1","FOXA1","CTCF"),FUN=function(z) plot_footprints(x=combined,footprints=z,outname="210924_cellline"),mc.cores=5) #plot and return 5 at a time

outname="210924_cellline"
pdf(paste0(outname,".motif_footprints.pdf"),height=5*length(out))
print(wrap_plots(out) + patchwork::plot_layout(ncol = 1))
dev.off()
system(paste("slack -F ",paste0(outname,".motif_footprints.pdf")," ryan_todo" ))

```

## Cistopic Correlation to TITAN (RNA Topic Modelling)
Titan was run by Aaron Doe on the RNA data. Located here:
```bash
/home/groups/CEDAR/doe/projects/ATACRNA
```

```R
library(Signac)
library(Seurat)
library(ComplexHeatmap)
library(JASPAR2020)
library(TFBSTools)
library(grid)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(parallel)
  library(BSgenome.Hsapiens.UCSC.hg19)
  library(patchwork)

setwd("/home/groups/CEDAR/mulqueen/projects/10x_atacrna")
mcf7_rna<-readRDS("/home/groups/CEDAR/doe/projects/ATACRNA/MCF7_RNA_strictQC.rds")
t47d_rna<-readRDS("/home/groups/CEDAR/doe/projects/ATACRNA/T47D_RNA_strictQC.rds")

mcf7_bins<-readRDS("/home/groups/CEDAR/doe/projects/ATACRNA/MCF7_ATAC_wBins.rds")
t47d_bins<-readRDS("/home/groups/CEDAR/doe/projects/ATACRNA/T47D_ATAC_wBins.rds")

mcf7_atac<-readRDS("210924_mcf7.SeuratObject.Rds")
t47d_atac<-readRDS("210924_t47d.SeuratObject.Rds")

mcf7_bins$Topic_11bin
t47d_bins$Topic_11bin

#Correlate cistopic with titan topics
correlate_cistopic_titan<-function(x,y,outname){
cistopic_dat<-x@reductions$cistopic@cell.embeddings #get cistopic from x
row.names(cistopic_dat)<-substr(row.names(cistopic_dat),0,18)
titan_dat<-as.data.frame(y@reductions$imputedLDA@cell.embeddings) #get titan from y
cells<-row.names(titan_dat)[which(row.names(titan_dat) %in% row.names(cistopic_dat))] # set list of shared cell names
sum(cells %in% row.names(titan_dat))==sum(cells %in% row.names(titan_dat)) #check to make sure same number of cells
titan_dat<-titan_dat[cells,]#reorder data frames
cistopic_dat<-cistopic_dat[cells,]
cor_mat<-data.frame(matrix(ncol = ncol(cistopic_dat), nrow = ncol(titan_dat))) #make an empty data frame to populate
colnames(cor_mat)<-colnames(cistopic_dat)
row.names(cor_mat)<-colnames(titan_dat)

for (yi in 1:ncol(cistopic_dat)){
  for (xj in 1:ncol(titan_dat)){
    cor_mat[xj,yi]<-cor(x=as.numeric(cistopic_dat[,yi]),y=as.numeric(titan_dat[,xj]))
  }
}

out_plt<-Heatmap(cor_mat)
pdf(paste0(outname,".cistopic_titan_topiccorrelation.pdf"))
print(out_plt)
dev.off()
system(paste0("slack -F ",paste0(outname,".cistopic_titan_topiccorrelation.pdf")," ryan_todo"))

#add cistopic square correlation matrix
cor_cistopic<-cor(cistopic_dat)
out_plt<-Heatmap(cor_cistopic)
pdf(paste0(outname,".cistopic_squarecorrelation.pdf"))
print(out_plt)
dev.off()
system(paste0("slack -F ",paste0(outname,".cistopic_squarecorrelation.pdf")," ryan_todo"))

#titan square correlation
cor_titan<-cor(titan_dat)
out_plt<-Heatmap(cor_titan)
pdf(paste0(outname,".titan_squarecorrelation.pdf"))
print(out_plt)
dev.off()
system(paste0("slack -F ",paste0(outname,".titan_squarecorrelation.pdf")," ryan_todo"))

}

#x is atac library signac object with cistopic in reductions slot
#y is rna library seurat object with imputedLDA in reductions slot
correlate_cistopic_titan(x=mcf7_atac,y=mcf7_rna,outname="mcf7")
correlate_cistopic_titan(x=t47d_atac,y=t47d_rna,outname="t47d")


#now to correlate chromvar motifs with titan topics

#Correlate cistopic with titan topics
correlate_chromvar_titan<-function(x,y,outname){
chromvar_dat<-as.data.frame(t(x@assays$chromvar@data)) #get chromvar from x
row.names(chromvar_dat)<-substr(row.names(chromvar_dat),0,18)
tfList <- getMatrixByID(JASPAR2020, ID=colnames(chromvar_dat)) #set up readable chromvar names
tfList <-unlist(lapply(names(tfList), function(x) name(tfList[[x]])))
colnames(chromvar_dat)<-tfList

titan_dat<-as.data.frame(y@reductions$imputedLDA@cell.embeddings) #get titan from y
cells<-row.names(titan_dat)[which(row.names(titan_dat) %in% row.names(chromvar_dat))] # set list of shared cell names
sum(cells %in% row.names(titan_dat))==sum(cells %in% row.names(titan_dat)) #check to make sure same number of cells
titan_dat<-titan_dat[cells,]#reorder data frames
chromvar_dat<-chromvar_dat[cells,]
cor_mat<-data.frame(matrix(ncol = ncol(chromvar_dat), nrow = ncol(titan_dat))) #make an empty data frame to populate
colnames(cor_mat)<-colnames(chromvar_dat)
row.names(cor_mat)<-colnames(titan_dat)

for (yi in 1:ncol(chromvar_dat)){
  for (xj in 1:ncol(titan_dat)){
    cor_mat[xj,yi]<-cor(x=chromvar_dat[,yi],y=titan_dat[,xj],use="complete.obs")
  }
}

out_plt<-Heatmap(t(cor_mat),
  row_names_gp = grid::gpar(fontsize = 4),
  row_km=5
  )
pdf(paste0(outname,".chromvar_titan_topiccorrelation.pdf"),height=30)
print(out_plt)
dev.off()
system(paste0("slack -F ",paste0(outname,".chromvar_titan_topiccorrelation.pdf")," ryan_todo"))
}

#x is atac library signac object with chromvar in assay slot
#y is rna library seurat object with imputedLDA in reductions slot
correlate_chromvar_titan(x=mcf7_atac,y=mcf7_rna,outname="mcf7")
correlate_chromvar_titan(x=t47d_atac,y=t47d_rna,outname="t47d")


#find DA motifs from top bins of topics
  #set up a named vector of topic11 and tpoic17 cells to add to seurat object
  topic_11<-setNames(mcf7_bins@meta.data$Topic_11bin,paste0(row.names(mcf7_bins@meta.data),mcf7_bins$origin))
  topic_11<-topic_11[topic_11 == "75percentile_Topic11"]
  topic_11_score<-setNames(mcf7_bins$ImputedTopic_11,paste0(row.names(mcf7_bins@meta.data),mcf7_bins$origin))
  topic_17<-setNames(mcf7_bins@meta.data$Topic_17bin,paste0(row.names(mcf7_bins@meta.data),mcf7_bins$origin))
  topic_17<-topic_17[topic_17 == "75percentile_Topic17"]
  topic_17_score<-setNames(mcf7_bins$ImputedTopic_17,paste0(row.names(mcf7_bins@meta.data),mcf7_bins$origin))
  mcf7_atac<-AddMetaData(object=mcf7_atac,metadata=topic_11,col.name="top_topic")
  mcf7_atac<-AddMetaData(object=mcf7_atac,metadata=topic_11_score,col.name="ImputedTopic_11")
  mcf7_atac<-AddMetaData(object=mcf7_atac,metadata=topic_17_score,col.name="ImputedTopic_17")
  levels(mcf7_atac$top_topic) <- c(levels(mcf7_atac$top_topic),"75percentile_Topic17")
  mcf7_atac@meta.data[row.names(mcf7_atac@meta.data) %in% names(topic_17),]$top_topic<-"75percentile_Topic17"
  mcf7_atac_subset<-subset(mcf7_atac,cells=which(is.finite(mcf7_atac$top_topic)))
  Idents(mcf7_atac_subset)<-mcf7_atac_subset$top_topic

  #and for t47d
  topic_11<-setNames(t47d_bins@meta.data$Topic_11bin,paste0(row.names(t47d_bins@meta.data),t47d_bins$origin))
  topic_11<-topic_11[topic_11 == "75percentile_Topic11"]
  topic_11_score<-setNames(t47d_bins$ImputedTopic_11,paste0(row.names(t47d_bins@meta.data),t47d_bins$origin))
  topic_17<-setNames(t47d_bins@meta.data$Topic_17bin,paste0(row.names(t47d_bins@meta.data),t47d_bins$origin))
  topic_17<-topic_17[topic_17 == "75percentile_Topic17"]
  topic_17_score<-setNames(t47d_bins$ImputedTopic_17,paste0(row.names(t47d_bins@meta.data),t47d_bins$origin))
  t47d_atac<-AddMetaData(object=t47d_atac,metadata=topic_11,col.name="top_topic")
  t47d_atac<-AddMetaData(object=t47d_atac,metadata=topic_11_score,col.name="ImputedTopic_11")
  t47d_atac<-AddMetaData(object=t47d_atac,metadata=topic_17_score,col.name="ImputedTopic_17")
  levels(t47d_atac$top_topic) <- c(levels(t47d_atac$top_topic),"75percentile_Topic17")
  t47d_atac@meta.data[row.names(t47d_atac@meta.data) %in% names(topic_17),]$top_topic<-"75percentile_Topic17"
  t47d_atac_subset<-subset(t47d_atac,cells=which(is.finite(t47d_atac$top_topic)))
  Idents(t47d_atac_subset)<-t47d_atac_subset$top_topic

#find differences in chromvar usage 
  mcf7_da_tf <- FindMarkers(object = mcf7_atac_subset, ident.1 = "75percentile_Topic11", ident.2="75percentile_Topic17", test.use = 'LR', latent.vars = "nCount_peaks", only.pos=F, assay="chromvar") 
  t47d_da_tf <- FindMarkers(object = t47d_atac_subset, ident.1 = "75percentile_Topic11", ident.2="75percentile_Topic17", test.use = 'LR', latent.vars = "nCount_peaks", only.pos=F, assay="chromvar") 

#set up readable motif names
  mcf7_da_tf$tf_name <- unlist(lapply(unlist(lapply(row.names(mcf7_da_tf), function(x) getMatrixByID(JASPAR2020,ID=x))),function(y) name(y)))
  t47d_da_tf$tf_name <- unlist(lapply(unlist(lapply(row.names(t47d_da_tf), function(x) getMatrixByID(JASPAR2020,ID=x))),function(y) name(y)))

#plot
  mcf7_da_tf$sig<-ifelse(mcf7_da_tf$p_val_adj<=0.01,"sig","non_sig")
  t47d_da_tf$sig<-ifelse(t47d_da_tf$p_val_adj<=0.01,"sig","non_sig")
  mcf7_da_tf$label<-""
  mcf7_da_tf[mcf7_da_tf$sig=="sig",]$label<-mcf7_da_tf[mcf7_da_tf$sig=="sig",]$tf_name
  t47d_da_tf$label<-""
  t47d_da_tf[t47d_da_tf$sig=="sig",]$label<-t47d_da_tf[t47d_da_tf$sig=="sig",]$tf_name

  plt<-ggplot(mcf7_da_tf,aes(x=avg_log2FC,y=-log10(p_val_adj),color=sig,label=label))+geom_point()+theme_bw()+scale_fill_manual(values=c("#999999", "#FF0000"))+geom_label_repel(label.size = 0.1,max.overlaps=20)
  ggsave(plt,file="mcf7_topic11v17_chromvar.pdf")
  system("slack -F mcf7_topic11v17_chromvar.pdf ryan_todo")

  plt<-ggplot(t47d_da_tf,aes(x=avg_log2FC,y=-log10(p_val_adj),color=sig,label=label))+geom_point()+theme_bw()+scale_fill_manual(values=c("#999999", "#FF0000"))+geom_label_repel(label.size = 0.1,max.overlaps=20)
  ggsave(plt,file="t47d_topic11v17_chromvar.pdf")
  system("slack -F t47d_topic11v17_chromvar.pdf ryan_todo")

#plot topic11 vs topic17 for cells
plt<-FeaturePlot(object = mcf7_atac, features = c("ImputedTopic_11", "ImputedTopic_17"),blend = T,col=c("lightgrey","red","blue"),reduction="umap",max.cutoff=10)
ggsave(plt,file="mcf7_topicimputation.umap.pdf",width=13)
system("slack -F mcf7_topicimputation.umap.pdf ryan_todo")

plt<-FeaturePlot(object = t47d_atac, features = c("ImputedTopic_11", "ImputedTopic_17"),blend = T,col=c("lightgrey","red","blue"),reduction="umap",max.cutoff=10)
ggsave(plt,file="t47d_topicimputation.umap.pdf",width=13)
system("slack -F t47d_topicimputation.umap.pdf ryan_todo")

#function for plotting footprints in data sets

plot_footprints<-function(x,footprints){
  x <- Footprint(object = x, motif.name = footprints, genome = BSgenome.Hsapiens.UCSC.hg19, in.peaks=T)   # gather the footprinting information for sets of motifs
  p2 <- PlotFootprint(x, features = footprints,label=F)   # plot the footprint data for each group of cells, might want to change idents
  return(p2)
}

Idents(t47d_atac_subset)<-t47d_atac_subset$top_topic
out1<-mclapply(c("ESR1","TP53","FOXA1","TEAD3"),FUN=function(z) plot_footprints(x=t47d_atac_subset,footprints=z),mc.cores=5) #plot and return 5 at a time
outname="210924_t47d_topimputedtopic"
pdf(paste0(outname,".motif_footprints.pdf"),height=5*length(out1))
print(wrap_plots(out1) + patchwork::plot_layout(ncol = 1))
dev.off()
system(paste("slack -F ",paste0(outname,".motif_footprints.pdf")," ryan_todo" ))

Idents(mcf7_atac_subset)<-mcf7_atac_subset$top_topic
out2<-mclapply(c("ESR1","FOXA1","TGIF1","MEF2A","ESR2"),FUN=function(z) plot_footprints(x=mcf7_atac_subset,footprints=z),mc.cores=5) #plot and return 5 at a time
outname="210924_mcf7_topimputedtopic"
pdf(paste0(outname,".motif_footprints.pdf"),height=5*length(out2))
print(wrap_plots(out2) + patchwork::plot_layout(ncol = 1))
dev.off()
system(paste("slack -F ",paste0(outname,".motif_footprints.pdf")," ryan_todo" ))


```

<!---
### Installing Cicero and Monocle3
I had a lot of trouble with the system gdal and my own conda gdal, so I restarted with a fresh conda environment.

```bash
# Create a new conda environment
conda create -n monocle
# Activate it
source activate monocle
# Install R and geospatial dependencies
conda install -c conda-forge r-essentials geos r-rgeos r-rgdal udunits2 #proj led to gdal error again
```
```R
#then install R packages
dyn.load('/home/groups/CEDAR/mulqueen/src/miniconda3/envs/monocle/lib/libgeos_c.so', local=FALSE) #https://github.com/r-spatial/sf/issues/844
install.pacakges("sf")
install.packages('raster')
install.packages('GISTools', dep=T)

if (!requireNamespace("BiocManager", quietly = TRUE))
install.packages("BiocManager")
BiocManager::install(version = "3.10")



BiocManager::install(c('BiocGenerics', 'DelayedArray','DelayedMatrixStats','limma', 'S4Vectors', 'SingleCellExperiment','SummarizedExperiment', 'batchelor','Matrix.utils'))

install.packages("devtools")
devtools::install_github('cole-trapnell-lab/leidenbase')
devtools::install_github('cole-trapnell-lab/monocle3')
BiocManager::install("cicero")

```
-->

## Cicero

```R
  library(Signac)
  library(Seurat)
  library(SeuratWrappers)
  library(ggplot2)
  library(patchwork)
  library(monocle3)
  library(cicero)
  library(SeuratObjects)
  library(EnsDb.Hsapiens.v75)
  setwd("/home/groups/CEDAR/mulqueen/projects/10x_atacrna")
  #Cicero processing function
  cicero_processing<-function(object_input,prefix){

      #Generate CDS format from Seurat object
      atac.cds <- as.CellDataSet(object_input,assay="peaks",reduction="umap")

      # convert to CellDataSet format and make the cicero object
      print("Making Cicero format CDS file")
      atac.cicero <- make_cicero_cds(atac.cds, reduced_coordinates = atac.cds@reducedDimS)
      saveRDS(atac.cicero,paste(prefix,"atac_cicero_cds.Rds",sep="_"))
      
      # extract gene annotations from EnsDb
      annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)

      # change to UCSC style since the data was mapped to hg38
      seqlevelsStyle(annotations) <- 'UCSC'
      genome(annotations) <- "hg19"

      genome <- annotations@seqinfo # get the chromosome sizes from the Seurat object
      genome.df <- data.frame("chr" = genome@seqnames, "length" = genome@seqlengths) # convert chromosome sizes to a dataframe
      
      print("Running Cicero to generate connections.")
      conns <- run_cicero(atac.cicero, genomic_coords = genome.df) # run cicero
      saveRDS(conns,paste(prefix,"atac_cicero_conns.Rds",sep="_"))
      
      print("Generating CCANs")
      ccans <- generate_ccans(conns) # generate ccans
      saveRDS(ccans,paste(prefix,"atac_cicero_ccans.Rds",sep="_"))
      
      print("Adding CCAN links into Seurat Object and Returning.")
      links <- ConnectionsToLinks(conns = conns, ccans = ccans) #Add connections back to Seurat object as links
      DefaultAssay(object_input)<-"peaks"
      Links(object_input) <- links
      return(object_input)
  }


combined<-readRDS("210924_cellline.SeuratObject.Rds")
combined<-cicero_processing(object_input=combined,prefix="210924_cellline")
saveRDS(combined,"210924_cellline.SeuratObject.unnormGA.Rds")

  # generate unnormalized gene activity matrix
  # gene annotation sample
  hg19_annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)

  pos <-as.data.frame(hg19_annotations,row.names=NULL)
  pos$chromosome<-paste0("chr",pos$seqnames)
  pos$gene<-pos$gene_id
  pos <- subset(pos, strand == "+")
  pos <- pos[order(pos$start),] 
  pos <- pos[!duplicated(pos$tx_id),] # remove all but the first exons per transcript
  pos$end <- pos$start + 1 # make a 1 base pair marker of the TSS

  neg <-as.data.frame(hg19_annotations,row.names=NULL)
  neg$chromosome<-paste0("chr",neg$seqnames)
  neg$gene<-neg$gene_id
  neg <- subset(neg, strand == "-")
  neg <- neg[order(neg$start,decreasing=TRUE),] 
  neg <- neg[!duplicated(neg$tx_id),] # remove all but the first exons per transcript
  neg$end <- neg$end + 1 # make a 1 base pair marker of the TSS

  gene_annotation<- rbind(pos, neg)
  gene_annotation <- gene_annotation[,c("chromosome","start","end","gene_name")] # Make a subset of the TSS annotation columns containing just the coordinates and the gene name
  names(gene_annotation)[4] <- "gene" # Rename the gene symbol column to "gene"

  geneactivity_processing<-function(cds_input,conns_input,prefix){
      atac.cds<- annotate_cds_by_site(cds_input, gene_annotation)
      unnorm_ga <- build_gene_activity_matrix(atac.cds, conns_input)
      saveRDS(unnorm_ga,paste(prefix,"unnorm_GA.Rds",sep="."))
  }

  conns<-as.data.frame(readRDS("210924_cellline_atac_cicero_conns.Rds"))
  obj.cicero<-readRDS("210924_cellline_atac_cicero_cds.Rds")
  geneactivity_processing(cds_input=as.CellDataSet(combined,assay="peaks",reduction="umap"),conns_input=conns,prefix="210924_cellline")

  #Read in unnormalized GA
  cicero_gene_activities<-readRDS("210924_cellline.unnorm_GA.Rds")
  combined[['GeneActivity']]<- CreateAssayObject(counts = cicero_gene_activities) 

  # normalize
  combined <- NormalizeData(
    object = combined,
    assay = 'GeneActivity',
    normalization.method = 'LogNormalize',
    scale.factor = median(combined$nCount_GeneActivity)
  )
  saveRDS(combined,"210924_cellline.SeuratObject.Rds")

```


